src/telemetry/transport/io.rs

```diff
@@ -0,0 +1,159 @@
+use bytes::Buf;
+use futures::{Async, Future, Poll};
+use std::io;
+use tokio_connect;
+use tokio::io::{AsyncRead, AsyncWrite};
+
+use transport::{Connection, Peek};
+
+use super::{NewSensor, Sensor, Eos};
+
+/// Wraps a transport with telemetry.
+#[derive(Debug)]
+pub struct Io<T> {
+    io: T,
+    sensor: Sensor,
+}
+
+/// Builds client transports with telemetry.
+#[derive(Clone, Debug)]
+pub struct Connect<C> {
+    underlying: C,
+    new_sensor: NewSensor,
+}
+
+/// Adds telemetry to a pending client transport.
+#[derive(Clone, Debug)]
+pub struct Connecting<C: tokio_connect::Connect> {
+    underlying: C::Future,
+    new_sensor: Option<NewSensor>,
+}
+
+// === impl Io ===
+
+impl<T: AsyncRead + AsyncWrite> Io<T> {
+    pub(super) fn new(io: T, sensor: Sensor) -> Self {
+        Self { io, sensor }
+    }
+
+    /// Wraps an operation on the underlying transport with error telemetry.
+    ///
+    /// If the transport operation results in a non-recoverable error, record a
+    /// transport closure.
+    fn sense_err<F, U>(&mut self, op: F) -> io::Result<U>
+    where
+        F: FnOnce(&mut T) -> io::Result<U>,
+    {
+        match op(&mut self.io) {
+            Ok(v) => Ok(v),
+            Err(e) => {
+                if e.kind() != io::ErrorKind::WouldBlock {
```

Comment:
TIOLI: What do you think of
```rust
match op(&mut self.io) {
    Ok(v) => Ok(v),
    Err(e) if e.kind() != io::ErrorKind::WouldBlock => {
        let eos = Eos::Error { errno: e.raw_os_error().map(|e| e.into()) };
        self.sensor.record_close(eos);
        Err(e)
    },
    e => e,
```

Reply:
that seems fine, but i don't think it's really better/worse than what's there now.

Reply:
i felt like it made the control flow _slightly_ easier to follow, but 🤷‍♀️